waku/v2/utils/logger.go

```diff
@@ -15,33 +17,39 @@ func SetLogLevel(level string) error {
 		return err
 	}
 	atom.SetLevel(lvl)
+
 	return nil
 }
 
 func Logger() *zap.Logger {
 	if log == nil {
-		cfg := zap.Config{
-			Encoding:         "console",
-			Level:            atom,
-			OutputPaths:      []string{"stderr"},
-			ErrorOutputPaths: []string{"stderr"},
-			EncoderConfig: zapcore.EncoderConfig{
-				MessageKey:   "message",
-				LevelKey:     "level",
-				EncodeLevel:  zapcore.CapitalLevelEncoder,
-				TimeKey:      "time",
-				EncodeTime:   zapcore.ISO8601TimeEncoder,
-				NameKey:      "caller",
-				EncodeCaller: zapcore.ShortCallerEncoder,
-			},
-		}
-
-		logger, err := cfg.Build()
-		if err != nil {
-			panic("could not create logger")
-		}
-
-		log = logger.Named("gowaku")
+		InitLogger("console").Panic("Logger not yet initialized")
+	}
+	return log
+}
+
+func InitLogger(logEncoding string) *zap.Logger {
```

Comment:
How about something like this?
```go
func JSONLogger() *zap.Logger {
  return LoggerWithEncoding("json")
}

func Logger() *zap.Logger {
  return LoggerWithEncoding("console")
}

func LoggerWithEncoding(logEncoding string) *zap.Logger {
	if log == nil {
		cfg := zap.Config{
			Encoding:         logEncoding,
			Level:            atom,
			OutputPaths:      []string{"stderr"},
			ErrorOutputPaths: []string{"stderr"},
			EncoderConfig: zapcore.EncoderConfig{
				MessageKey:   "message",
				LevelKey:     "level",
				EncodeLevel:  zapcore.CapitalLevelEncoder,
				TimeKey:      "time",
				EncodeTime:   zapcore.ISO8601TimeEncoder,
				NameKey:      "caller",
				EncodeCaller: zapcore.ShortCallerEncoder,
			},
		}

		logger, err := cfg.Build()
		if err != nil {
			panic("could not create logger")
		}

		log = logger.Named("gowaku")
		InitLogger("console").Panic("Logger not yet initialized")
	}
	return log
}
```

Reply:
Would work. It just feels a little weird to be calling `LoggerWithEncoding("console")` every time you want to get a new logger (they call `utils.Logger()` all over the place), only to get back a JSON encoded logger. Seems a little unintuitive, since the first caller would always get to set the value with the `if log == nil {...` and all subsequent calls would have to take the existing logger.

Reply:
I do take your point that needing the initialization step is a little awkward too.

Reply:
I think normally just `Logger` would continue to be used as-is, and `LoggerWithEncoding` would only be used when you want to pass through the flag. `JSONLogger` actually probably not needed since it'd never be called in code anyway. That leaves the already package-public `Logger` method as-is, but adds an initializer that supports configuring the encoding too.

Reply:
Right. I think the weird thing is that with the code as outlined above, whichever code path is the first to call `LoggerWithEncoding` will set the encoding for all future loggers, since it only changes the stored value if `logger == nil`.

Right now, that happens to be `waku/v2/node/wakuoptions.go:83` because that's the first logger that gets called during app init.

Reply:
Oh hmm, yea that's super weird, I didn't see it setting the `log` outside of the method there. Now that you point that out, `InitLogger` makes more sense as a name.

Reply:
That came from the original status code. Looks like a perf optimization to not have to reinstantiate the logger every time someone calls the function.

Reply:
How about adding a `utils.NewLoggerWithEncoding(encoding string)` or just `utils.NewLogger(encoding string)` and use that instead `utils.Logger`, where the `New` version would just return a new instance of it every time, which is only when the node is initialized/created right?

Reply:
So, the unfortunate thing about the way the logger is being used in the app is that there are 11 other instances where they call `utils.Logger()` where the configuration info isn't readily available. For example, here: https://github.com/xmtp/go-waku/blob/nmolnar/improve-logging/waku/v2/metrics/metrics.go#L78

Reply:
This would all be nice and straightforward if all usage of the logger was based off of the root instance created at node startup. But they are relying on that `log` pointer to get a well-configured logger ad-hoc.